src/peft/tuners/xlora/classifier.py

```diff
@@ -0,0 +1,257 @@
+        # scalings = [batch_size, seq_len, n_layers, n_classes]
+
+        if self.config.enable_softmax:
+            scalings = self.softmax(scalings)
+
+        if self.n_predictions_lifetime > 0:
+            print(f"Scaling predictions: {scalings}")
+            self.n_predictions_lifetime -= 1
+
+        if self.scalings_logging:
+            self.log_scalings.append(scalings)
+
+        return scalings
+
+    def get_nb_trainable_parameters(self):
+        # https://github.com/huggingface/peft/blob/main/src/peft/mixed_model.py#L156
+        r"""
+        Returns the number of trainable parameters and number of all parameters in the model.
+        """
+        trainable_params = 0
+        all_param = 0
+        for _, param in self.named_parameters():
+            num_params = param.numel()
+            # if using DS Zero 3 and the weights are initialized empty
+            if num_params == 0 and hasattr(param, "ds_numel"):
+                num_params = param.ds_numel  # type: ignore
+
+            # Due to the design of 4bit linear layers from bitsandbytes
+            # one needs to multiply the number of parameters by 2 to get
+            # the correct number of parameters
+            if param.__class__.__name__ == "Params4bit":
+                num_params = num_params * 2
+
+            all_param += num_params
+            if param.requires_grad:
+                trainable_params += num_params
+
+        return trainable_params, all_param
+
+    @staticmethod
+    def _save_scalings(file: str, scalings: List[torch.Tensor]):
+        result = torch.cat(scalings, dim=0)
+        npy = result.numpy()
+        numpy.save(file, npy)
+
+    def flush_log_scalings(self, path: str):
+        if not self.scalings_logging:
+            raise Exception("Scalings logging is disabled!")
+
+        if len(self.log_scalings) == 0:
+            raise ValueError("No log scalings to flush.")
+
+        seqlens_map: Dict[int, Tuple[List[int], List[torch.Tensor]]] = {}
+        for i, scaling in enumerate(self.log_scalings):
+            seq_len = scaling.shape[0]
+            if seq_len not in seqlens_map:
+                seqlens_map[seq_len] = ([i], [scaling])
+            else:
+                seqlens_map[seq_len][0].append(i)
+                seqlens_map[seq_len][1].append(scaling)
+
+        if len(seqlens_map) == 1:
+            self._save_scalings(path, [scaling.unsqueeze(0) for scaling in self.log_scalings])
+        else:
+            indices_map: Dict[str, List[int]] = {}
+            for seq_len, (indices, scalings_list) in seqlens_map.items():
+                indices_map[f"{path}-{seq_len}.npy"] = indices
+
+                self._save_scalings(f"{path}-{seq_len}", [scaling.unsqueeze(0) for scaling in scalings_list])
+
+            with open(f"{path}-mapping.json", "w") as f:
+                f.write(json.dumps(indices_map))
+
+        self.log_scalings = []
```

Comment:
How about `self.log_scalings.clear()` in case someone else holds a reference to it?

Reply:
In `get_scalings_log` where I return the log, I make a shallow copy. So (unless somebody manually read the attribute), as an invariant, we should have the only reference.

Reply:
Clearing would work, there is no disadvantage. I think that leaving the list would take up too much memory, as all the scalings tensors (which can be large) would be kept. I have changed it to using `.clear()`.